staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go

```diff
@@ -981,6 +1026,61 @@ func ValidateCustomResourceDefinitionOpenAPISchema(schema *apiextensions.JSONSch
 	return allErrs
 }
 
+func extractMaxElements(schema *apiextensions.JSONSchemaProps) *uint64 {
+	switch schema.Type {
+	case "object":
+		if schema.AdditionalProperties != nil {
+			if schema.MaxProperties != nil {
+				maxProps := uint64(*schema.MaxProperties)
```

Comment:
What if user put a negative number in MaxProperties/MaxItems. As far as I aware of, we didn't check so user could potentially add negative under maxLength/maxItems/maxProperties. cc @jpbetz

Reply:
The same concern applied to the following. Worth to add a check

Reply:
that's a good point. I'm not sure we require those to be >= 0. when computing cost, we should bound to a minimum of 0 places like this as well:

```
maxProperties = *s.ValueValidation.MaxProperties
maxItems = *s.ValueValidation.MaxItems
byteWithMaxLength.MaxElements = *s.ValueValidation.MaxLength
strWithMaxLength.MaxElements = *s.ValueValidation.MaxLength * 4
```

Reply:
Even if we had validation that Max{Items/Properties/Length} >= 0, it would happen in parallel with the CEL validation.  I suppose we could refuse to validate any CEL rules if there are any validation errors related to Max{Items/Properties/Length} (similar to what @cici37 has done in CR validation, but in this case for CRD validation).  Adding validation rules to Max{Items/Properties/Length} would be API breaking though.

If MaxProperties is negative here, the length is going to be super high (overflow) and the cost is going to be super high. It's certainly more correct to do a `if 0`. I'm not particularly worried about it because if the user does set a negative value they just end up getting penalized with a really high CEL cost.

Reply:
let's consistently bound to a min of 0 with all the places we use schema-specified max length/items/properties

Reply:
SGTM since it looks easy enough to add.

Reply:
Agree with bounding min of 0. We could not do the way of refusing to validate any CEL rules if there are any validation errors related to Max{Items/Properties/Length} like what I did in CR validation because we don't(and should not) get Max{Items/Properties/Length} validation err during CRD creation.